framework/src/org/checkerframework/framework/type/AnnotatedTypeLoader.java

```diff
@@ -0,0 +1,408 @@
+ * providing an override implementation of
+ * {@link #isSupportedAnnotationClass(Class)}. See
+ * {@link org.checkerframework.checker.units.UnitsAnnotatedTypeLoader
+ * UnitsAnnotatedTypeLoader} for an example.
+ *
+ * @author Jeff Luo
+ */
+public class AnnotatedTypeLoader {
+    // For issuing errors to the user
+    private BaseTypeChecker checker;
+
+    // For loading from a source package directory
+    private String packageName;
+    private static final String QUAL_PACKAGE_SUFFIX = ".qual";
+
+    // For loading from a Jar file
+    private static final String CLASS_SUFFIX = ".class";
+
+    // For loading from external directories
+    private static final String JAVA_SUFFIX = ".java";
+
+    // Constants
+    private static final char DOT = '.';
+    private static final char SLASH = '/';
+
+    /**
+     * Processing Env used to create an {@link AnnotationBuilder}, which is in
+     * turn used to build the annotation mirror from the loaded class.
+     */
+    protected ProcessingEnvironment processingEnv;
+
+    // Stores the resource URL of the qual directory of a checker class
+    private URL resourceURL;
+
+    // Stores set of the loaded annotation classes
+    private Set<Class<? extends Annotation>> loadedAnnotations;
+
+    // Boolean flag to ensure we load only once
+    private boolean classesLoaded;
+
+    /**
+     * Constructor for loading annotations defined for a checker.
+     *
+     * @param checker
+     *            a {@link BaseTypeChecker} or its subclass
+     */
+    public AnnotatedTypeLoader(BaseTypeChecker checker) {
+        this.checker = checker;
+        processingEnv = checker.getProcessingEnvironment();
+        classesLoaded = false;
+
+        // package name must use dots, this is later prepended to annotation
+        // class names as we load the classes using the class loader
+        packageName = checker.getClass().getPackage().getName() + QUAL_PACKAGE_SUFFIX;
+
+        // Sees if the package name of the qual directory is a valid resource
+        // and retrieve its resource URL (either a jar or regular file directory)
+        // In checkers, there will be a resource URL for the qual directory. But
+        // when called in the framework (eg GeneralAnnotatedTypeFactory), there
+        // won't be a resourceURL since there isn't a qual directory
+
+        // resource URLs must use slashes
+        resourceURL = this.getClass().getClassLoader().getResource(packageName.replace(DOT, SLASH));
+        // resourceURL = Thread.currentThread().getContextClassLoader().getResource(packageName.replace(DOT, SLASH));
+
+        loadedAnnotations = new HashSet<Class<? extends Annotation>>();
+    }
+
+    /**
+     * Gets the set of the loaded annotation classes.
+     *
+     * @return the set of loaded annotation classes
+     */
+    public final Set<Class<? extends Annotation>> getLoadedAnnotationClasses() {
+        return loadedAnnotations;
```

Comment:
You should just document what sharing might exist. If the loadedAnnotations Set gets changed by the ATF, that change will be visible to other users of that Loader. So this interaction between the two objects needs some documentation.